packages/vat-data/test/test-vivify.js

```diff
@@ -0,0 +1,121 @@
+// @ts-check
+
+// Modeled on test-heap-classes.js
+
+import { test } from '@agoric/swingset-vat/tools/prepare-test-env-ava.js';
+import { M } from '@agoric/store';
+import {
+  vivifyFarClass,
+  vivifyFarClassKit,
+  vivifyFarInstance,
+} from '../src/far-class-utils.js';
+import { makeScalarBigMapStore } from '../src/vat-data-bindings.js';
+
+const UpCounterI = M.interface('UpCounter', {
+  incr: M.call()
+    // TODO M.number() should not be needed to get a better error message
+    .optional(M.and(M.number(), M.gte(0)))
+    .returns(M.number()),
+});
+
+const DownCounterI = M.interface('DownCounter', {
+  decr: M.call()
+    // TODO M.number() should not be needed to get a better error message
+    .optional(M.and(M.number(), M.gte(0)))
+    .returns(M.number()),
+});
+
+test('test vivifyFarClass', t => {
+  const baggage = makeScalarBigMapStore('baggage', { durable: true });
+
+  const makeUpCounter = vivifyFarClass(
+    baggage,
+    'UpCounter',
+    UpCounterI,
+    (x = 0) => ({ x }),
+    {
+      incr(y = 1) {
+        // @ts-expect-error TS doesn't know that `this` is a `Context`
+        const { state } = this;
+        state.x += y;
+        return state.x;
+      },
+    },
+  );
+  const upCounter = makeUpCounter(3);
+  t.is(upCounter.incr(5), 8);
+  t.is(upCounter.incr(1), 9);
+  t.throws(() => upCounter.incr(-3), {
+    message: 'In "incr" method of (UpCounter) arg 0: -3 - Must be >= 0',
+  });
+  // @ts-expect-error TS doesn't know that `this` is a `Context`
+  t.throws(() => upCounter.incr('foo'), {
+    message:
+      'In "incr" method of (UpCounter) arg 0: string "foo" - Must be a number',
+  });
+});
+
+test('test vivifyFarClassKit', t => {
+  const baggage = makeScalarBigMapStore('baggage', { durable: true });
+
+  const makeCounterKit = vivifyFarClassKit(
+    baggage,
+    'Counter',
+    harden({ up: UpCounterI, down: DownCounterI }),
```

Comment:
This is a case where we really should not require `harden`. We are calling a library that is essentially a part of the computational model, and the `harden` obscures the most important thing here.

Reply:
Another good catch. Not only do I agree, but my style is not to use harden in this positions. It would indeed be appropriate for a static checker to know all of these definers as hardeners, the way it would know `harden` and `Far` are hardeners.

Reply:
Oh, I didn't get at first what's different. This is the interfaceKit argument, not the methodsKit argument. The methods and methodsKit are indeed not frozen.

So I still agree, but at least I had not fallen out of the style I'm trying to practice.

Reply:
Done